File: 03-week-3--additional-practice/00-bonus-1--balancing-parenetheses/javascript/balancing_parentheses.js
function balancingParentheses(string) {
  // type your code here
  let result = 0
  let paranthesesO = 0
  let paranthesesC = 0

  const stringArr = string.split('')

  stringArr.forEach(el => {
    if (el === '(') {
      paranthesesO += 1
    }
    if (el === ')') {
      paranthesesC += 1
    }
    })
    if (paranthesesC === paranthesesO) {
      return 0
    }
    if (paranthesesC  > paranthesesO) {
      result = paranthesesC - paranthesesO
    }
    if (paranthesesC < paranthesesO) {
      result = paranthesesO - paranthesesC
    }
      return result
} 

function balancingParenthesesSol(string) {
  let missing = 0;
  let openings = 0;

  for (let i = 0; i < string.length; ++i) {
    if (string[i] === '(') {
      ++openings;
      continue;
    }

    if (openings > 0) {
      --openings;
    } else {
      ++missing;
    }
  }

  return missing + openings;
}

// const balancingParenthesesRecursive = (string) => {

//   const stringArr = string.split('')
//   let result = 0
//   let paranthesesO = 0
//   let paranthesesC = 0

//   if (stringArr.length === 0 ) {
//     return 
//   }
//   console.log(stringArr)
//   if (stringArr[0] === '(') {
//     paranthesesO += 1
//   }
//   if (stringArr[0] === ')') {
//     paranthesesC += 1
//   }

//   balancingParenthesesRecursive(stringArr[0])

//   if (paranthesesC === paranthesesO) {
//     return 0
//   }
//   if (paranthesesC  > paranthesesO) {
//     result = paranthesesC - paranthesesO
//   }
//   if (paranthesesC < paranthesesO) {
//     result = paranthesesO - paranthesesC
//   }
//   return result
// }

balancingParenthesesRecursive('(()())))(())((')

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log(balancingParentheses('(()())'));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses('()))'));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses(')'));

  console.log("Expecting: 0");
  console.log(balancingParentheses('(()())))(())(('));

  console.log("");

  console.log("Expecting: 3");
  console.log(balancingParentheses('())))'));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses('(()()))'));

  console.log("Expecting: 0");
  console.log(balancingParentheses(''));
}

function benchmark(callback) {
  const startTime = Date.now();

  for (let i = 0; i < 1000; ++i) {
    callback();
  }

  return (Date.now() - startTime) / 1000;
}

console.log("iterative", benchmark(() => balancingParentheses(')())(())((()(()(')))
console.log("iterative For Loop", benchmark(() => balancingParenthesesSol(')())(())((()(()(')))

module.exports = balancingParentheses;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 03-week-3--additional-practice/00-bonus-1--balancing-parenetheses/ruby/balancing_parentheses.rb
def balancing_parentheses(string)
  # type your code in here

  parathesesO = 0
  parathesesC = 0

  string.chars.each do |char|
    if char == '('
      parathesesO += 1
      next
    end
    
    if parathesesO > 0 
      parathesesO -= 1 
    else
      parathesesC += 1
    end
  end
  parathesesC + parathesesO
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts balancing_parentheses('(()())')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses('()))')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses(')')

  puts

  puts "Expecting: 0"
  puts balancing_parentheses('()')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses('(')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses(')(')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses(')()')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses(')((((()))((())))()(()))(')

  puts

  puts "Expecting: 3"
  puts balancing_parentheses(')))')

  puts

  puts "Expecting: 3"
  puts balancing_parentheses('(((')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
